### PR TITLE
OSD: Fix check_past_interval_bounds()

### PR DESCRIPTION
**Pull request by Matan-B**
* [0cb809cd5f368367bad3de11e55fa4cf8a228865]: When getting the required past interval bounds we use oldest_map or current pg info (lec/ec).
Before this change we set oldest_map epoch using the osd's superblock.oldest_map.
The fix will use the max_oldest_map(9789c29fe0f82b236703f7ae4d356251fde3a46f) received with other peers instead,
since a specific osd's oldest_map can lag for a while in order to avoid large workloads. (see example below)

* [2aa759f2a5c4c33f60da70dd46d4acd28bb285b3]: max_oldest_map is moved from OSDService::max_oldest_map into OSDSuperblock for persistence.
* [47e988cb28d081af85156bbf9f9457f861213c91]: oldest_stored_osdmap() was removed.  The only usage was for identifying map gaps on new intervals.
    We should use max_oldest_stored_osdmap() instead, since a specific osd's oldest_map might lag behind for a while.

* [0b6cc4d3106abf48329b31f5b01acead58e8b64b]: Past Intervals Documentation added.

Fixes: https://tracker.ceph.com/issues/49689

### Failure example (osd.11):
Assertion:
```
osd.11 pg_epoch: 375 pg[3.11s0( empty .. les/c/f=115/115/0 .. pi=[362,375)..] 3.11s0 
past_intervals [362,375) start interval does not contain the required bound [151,375) start
```
required past interval bounds are  [**151**,375) set from osd's superblock.oldest_map which is **151**:

```
osd.11 374 handle_osd_map epochs [375,375], i have 374, src has [362,375]
osd.11 374  removing old osdmap epoch 121
...
osd.11 374  removing old osdmap epoch 150
osd.11 374 write_superblock sb(...e375 [151,375] lci=[18,375])
```
osd's superblock.oldest_map is 151 only because we stop trimming after removing 30 maps (``osd_target_transaction_size``).
`oldest_map` will be used later on to continue the removal from the last epoch removed (150 in this case).
`max_oldest_map` (362) is the relevant epoch we should consider when getting the required start bound instead.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by athanatos** [comment]
Ah, 9789c29fe0f82b236703f7ae4d356251fde3a46f did do part of what we need to do here by adding max_oldest_map.  However, we forget the value on OSD restart since it doesn't get persisted to the superblock.  Let's fix that in this PR.

**Review by anthonyeleven — APPROVED** [review]
Nice!  Docs LGTM; my approval does not cover the code, which someone else should review.

File: src/osd/osd_types.h
Comment (athanatos) [inline src/osd/osd_types.h:5426]: max_oldest_map = 0

File: src/osd/osd_types.h
Comment (athanatos) [inline src/osd/osd_types.h:5439]: Stray whitespace change

File: src/osd/osd_types.h
Comment (athanatos) [inline src/osd/osd_types.h:5443]:
```
             << " max oldest=" << sb.max_oldest_map << ")";
```

**Review by athanatos — CHANGES_REQUESTED** [review]
I think this is about right.  @neha-ojha Probably worth a second set of eyes?

**Comment by athanatos** [comment]
This does need to wait on being able run a rados suite.

**Review by neha-ojha — APPROVED** [review]
I think this change makes sense. Let's run it through rados and upgrade suites. This should bake in main for a while before we think of backporting it.

**Comment by ljflores** [comment]
Rados suite review: https://pulpito.ceph.com/?branch=wip-yuri-testing-2023-01-23-0926

Failures, unrelated:
    1. https://tracker.ceph.com/issues/58587 -- new tracker
    2. https://tracker.ceph.com/issues/58585
    3. https://tracker.ceph.com/issues/58098 -- fix merged to latest main
    4. https://tracker.ceph.com/issues/58256 -- fix merged to latest main
    5. https://tracker.ceph.com/issues/57900
    6. https://tracker.ceph.com/issues/58475
    7. https://tracker.ceph.com/issues/58560

Details:
    1. test_dedup_tool.sh: test_dedup_object fails when pool 'dedup_chunk_pool' does not exist - Ceph - RADOS
    2. rook: failed to pull kubelet image - Ceph - Orchestrator
    3. qa/workunits/rados/test_crash.sh: crashes are never posted - Ceph - RADOS
    4. ObjectStore/StoreTestSpecificAUSize.SpilloverTest/2: Expected: (logger->get(l_bluefs_slow_used_bytes)) >= (16 * 1024 * 1024), actual: 0 vs 16777216 - Ceph - RADOS
    5. mon/crush_ops.sh: mons out of quorum - Ceph - RADOS
    6. test_dashboard_e2e.sh: Conflicting peer dependency: postcss@8.4.21 - Ceph - Mgr - Dashboard
    7. test_envlibrados_for_rocksdb.sh failed to subscrib repo - Ceph

**Comment by ljflores** [comment]
This is just waiting on upgrade test approval. @athanatos can you approve?

File: src/crimson/osd/pg.h
Comment (athanatos) [inline src/crimson/osd/pg.h:366]: I should have commented on this before, but this naming is simply confusing -- max oldest stored osdmap doesn't mean much since there's only a single oldest stored osdmap on a particular osd.  Rename this to cluster_osdmap_trim_lower_bound().

File: src/osd/osd_types.h
Comment (athanatos) [inline src/osd/osd_types.h:5426]: cluster_osdmap_trim_lower_bound()

**Review by athanatos — CHANGES_REQUESTED** [review]
There's another missing piece.  When we send MOSDMap messages to other OSDs, we should populate the oldest_map field with cluster_osdmap_trim_lower_bound() rather than our local oldest_map.  MOSDMap::oldest_map should be renamed to MOSDMap::cluster_osdmap_trim_lower_bound.  This shouldn't have upgrade implications because an old OSD will decode it as oldest_map and its interpretation will still be correct.  @neha-ojha Does the above make sense to you?

**Comment by Matan-B** [comment]
> There's another missing piece. When we send MOSDMap messages to other OSDs, we should populate the oldest_map field with cluster_osdmap_trim_lower_bound() rather than our local oldest_map. MOSDMap::oldest_map should be renamed to MOSDMap::cluster_osdmap_trim_lower_bound. This shouldn't have upgrade implications because an old OSD will decode it as oldest_map and its interpretation will still be correct. @neha-ojha Does the above make sense to you?

I have actually pushed 526756f4cd022234bbd34418b905922b1448a658 as a DNM to have a discussion about it before seeing your comment.
I have noticed this as well while renaming `max_oldest` to `cluster_osdmap_trim_lower_bound`.

This change looks very much reasonable to me. I will leave the commits as is (until approved) for easier CR.

**Comment by Matan-B** [comment]
@neha-ojha, @athanatos - `MOSDMap::oldest_map` is renamed to `MOSDMap::cluster_osdmap_trim_lower_bound` and is populated from `OSDSuperblock::cluster_osdmap_trim_lower_bound` (instead of `OSDSuperblock::oldest_map`).

(I pushed these changes on top of the older commits for better context, I will reorder and arrange those together once reviewed)

Eventually the changes in this PR are:
* Moving `OSDService::max_oldest_map` into the `OSDSuperblock`.
* Make `PeeringState::check_past_interval_bounds()` use `OSDSuperblock::max_oldest_map` instead of `OSDSuperblock::oldest_map`.
* Rename `OSDSuperblock::max_oldest_map` to `OSDSuperblock::cluster_osdmap_trim_lower_bound`.
* Rename `MOSDMap::oldest_map` to `MOSDMap::cluster_osdmap_trim_lower_bound` and populate this field from the super block accordingly.

File: src/crimson/osd/osd.cc
Comment (athanatos) [inline src/crimson/osd/osd.cc:681]: Move tlb to just after the map ranges (just before the number of pgs).  Also, "tlb" -> "tlb: "

File: src/messages/MOSDMap.h
Comment (athanatos) [inline src/messages/MOSDMap.h:33]:
```
  /**
   * cluster_osdmap_trim_lower_bound
   *
   * Encodes a lower bound on the monitor's osdmap trim bound.  Recipients
   * can safely trim up to this bound.  The sender stores maps back to
   * cluster_osdmap_trim_lower_bound.
   * 
   * This field was formerly named oldest_map and encoded the oldest map
   * stored by the sender.  The primary usage of this field, however, was to
   * allow the recipient to trim.  The secondary usage was to inform the
   * recipient of how many maps the sender stored in case it needed to request
   * more.  For both purposes, it should be safe for an older OSD to interpret
   * this field as oldest_map, and it should be safe for a new osd to interpret
   * the oldest_map field sent by an older osd as
   * cluster_osdmap_trim_lower_bound.  See bug <insert_tracker_url_here>
   */

File: src/osd/OSD.cc
Comment (athanatos) [inline src/osd/OSD.cc:1421]: I didn't check, but only the first bit of this line should be in the rename commit.  The rename commit should only have the m->m->oldest_map -> m->cluster_osdmap_trim_lower_bound bit.  The change from '= sblock.oldest_map' to '= sblock.cluster_osdmap_trim_lower_bound' should be in the commit introducing sblock.cluster_osdmap_trim_lower_bound.  It's annoying, but rename commits really shouldn't change behavior at all.

File: src/osd/PeeringState.cc
Comment (athanatos) [inline src/osd/PeeringState.cc:934]: Let's replace this comment with:

```
  // cluster_osdmap_trim_lower_bound gives us a bound on needed
  // intervals, see doc/dev/osd_internals/past_intervals.rst
```

File: src/osd/osd_types.h
Comment (athanatos) [inline src/osd/osd_types.h:5450]: Just above, we do " lci=[", so change this to " tlb="

**Review by athanatos — CHANGES_REQUESTED** [review]
Close!  I think this is right other than my nits.  Make sure that renaming MOSDMap::oldest_map to MOSDMap::cluster_osdmap_trim_lower_bound is in its own commit that changes literally nothing else -- it's useful for later readers of these commits to be able to simply skip that commit if they can trust it won't have any functional changes.

File: src/osd/OSD.cc
Comment (Matan-B) [reply src/osd/OSD.cc:1421]: > m->oldest_map -> m->cluster_osdmap_trim_lower_bound bit

* 721c9aea1869bf95a167b8879756b4f71fbf41bf

>   '= sblock.oldest_map' to '= sblock.cluster_osdmap_trim_lower_bound'

* ff58b81ddff5e99a57ca6cf2c5dae560fbc8adad

**Comment by Matan-B** [comment]
> Close! I think this is right other than my nits. Make sure that renaming MOSDMap::oldest_map to MOSDMap::cluster_osdmap_trim_lower_bound is in its own commit that changes literally nothing else -- it's useful for later readers of these commits to be able to simply skip that commit if they can trust it won't have any functional changes.

Thanks! The nits are addressed. Renaming commits of OSDSuperblock and MOSDMap are in their own commits.

**Review by athanatos — APPROVED** [review]

**Comment by ljflores** [comment]
Rados suite review: https://pulpito.ceph.com/?branch=wip-yuri6-testing-2023-02-24-0805

Failures, unrelated:
    1. https://tracker.ceph.com/issues/58585
    2. https://tracker.ceph.com/issues/58560
    3. https://tracker.ceph.com/issues/58797
    4. https://tracker.ceph.com/issues/58744
    5. https://tracker.ceph.com/issues/58475

Details:
    1. rook: failed to pull kubelet image - Ceph - Orchestrator
    2. test_envlibrados_for_rocksdb.sh failed to subscribe to repo - Infrastructure
    3. scrub/osd-scrub-dump.sh: TEST_recover_unexpected fails from "ERROR: Unexpectedly low amount of scrub reservations seen during test" - Ceph - RADOS
    4. qa: intermittent nfs test failures at nfs cluster creation - Ceph - CephFS
    5. test_dashboard_e2e.sh: Conflicting peer dependency: postcss@8.4.21 - Ceph - Mgr - Dashboard

**Pull request merged**
